File: cudf-sys/cpp/table.cpp
#include "table.hpp"

#include <cudf/column/column.hpp>
#include <cudf/column/column_view.hpp>
#include <cudf/table/table.hpp>
#include <cudf/table/table_view.hpp>

namespace cudf_rs {

Table::Table(std::unique_ptr<cudf::table> table) : table_(std::move(table)) {}

Table::~Table() = default;

std::unique_ptr<Table> Table::empty() noexcept {
  std::vector<std::unique_ptr<cudf::column>> empty(0);
  std::unique_ptr<cudf::table> table =
      std::make_unique<cudf::table>(std::move(empty));
  return std::make_unique<Table>(std::move(table));
}

cudf::size_type Table::alloc_size() const { return table_->alloc_size(); };

cudf::column const &Table::get_column(cudf::size_type column_index) const {
  return table_->get_column(column_index);
}

cudf::column_view const &
Table::get_column_view(cudf::size_type column_index) const {
  return this->view().column(column_index);
}

cudf::size_type Table::num_columns() const noexcept {
  return table_->num_columns();
};

cudf::size_type Table::num_rows() const noexcept { return table_->num_rows(); };

cudf::table_view const &Table::view() const {
  std::call_once(table_view_once_, [this] {
    table_view_ = std::make_unique<cudf::table_view>(table_->view());
  });
  return *table_view_;
}

} // namespace cudf_rs

File: cudf-sys/src/table.rs
#[cxx::bridge]
mod ffi {
    #[namespace = "cudf"]
    unsafe extern "C++" {
        type column = crate::column::column;
        type column_view = crate::column::column_view;
        type table_view;
    }

    #[namespace = "cudf_rs"]
    unsafe extern "C++" {
        include!("cudf-sys/include/table.hpp");

        type Table;

        #[Self = "Table"]
        fn empty() -> UniquePtr<Table>;

        fn alloc_size(self: &Table) -> Result<i32>;
        fn get_column(self: &Table, column_index: i32) -> Result<&column>;
        fn get_column_view(self: &Table, column_index: i32) -> Result<&column_view>;
        fn num_columns(self: &Table) -> i32;
        fn num_rows(self: &Table) -> i32;
        fn view(self: &Table) -> Result<&table_view>;
    }
}

pub use ffi::*;

unsafe impl Send for ffi::Table {}
unsafe impl Sync for ffi::Table {}
